engine/src/flutter/shell/platform/windows/window_manager.h

```diff
@@ -0,0 +1,138 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_WINDOWS_WINDOW_MANAGER_H_
+#define FLUTTER_SHELL_PLATFORM_WINDOWS_WINDOW_MANAGER_H_
+
+#include <windows.h>
+#include <optional>
+#include <unordered_map>
+#include <vector>
+
+#include "flutter/shell/platform/common/public/flutter_export.h"
+
+#include "flutter/fml/macros.h"
+#include "flutter/shell/platform/common/isolate_scope.h"
+#include "flutter/shell/platform/embedder/embedder.h"
+
+namespace flutter {
+
+class FlutterWindowsEngine;
+class HostWindow;
+struct WindowingInitRequest;
+
+struct WindowsMessage {
+  FlutterViewId view_id;
+  HWND hwnd;
+  UINT message;
+  WPARAM wParam;
+  LPARAM lParam;
+  LRESULT result;
+  bool handled;
+};
+
+struct WindowSizing {
+  bool has_size;
+  double width;
+  double height;
+  bool has_constraints;
+  double min_width;
+  double min_height;
+  double max_width;
+  double max_height;
+};
+
+struct WindowingInitRequest {
+  void (*on_message)(WindowsMessage*);
+};
+
+struct WindowCreationRequest {
+  WindowSizing content_size;
+};
+
+// A manager class for managing |HostWindow| instances.
+// A unique instance of this class is owned by |FlutterWindowsEngine|.
+class WindowManager {
+ public:
+  explicit WindowManager(FlutterWindowsEngine* engine);
+  virtual ~WindowManager() = default;
+
+  void Initialize(const WindowingInitRequest* request);
+
+  bool HasTopLevelWindows() const;
+
+  FlutterViewId CreateRegularWindow(const WindowCreationRequest* request);
+
+  // Message handler called by |HostWindow::WndProc| to process window
+  // messages before delegating them to the host window. This allows the
+  // controller to process messages that affect the state of other host windows.
+  std::optional<LRESULT> HandleMessage(HWND hwnd,
+                                       UINT message,
+                                       WPARAM wparam,
+                                       LPARAM lparam);
+
+  void OnEngineShutdown();
+
+ private:
+  // The Flutter engine that owns this controller.
+  FlutterWindowsEngine* const engine_;
+
+  // Callback that relays windows messages to the isolate. Set
+  // during Initialize().
+  void (*on_message_)(WindowsMessage*) = nullptr;
+
+  // Isolate that runs the Dart code. Set during Initialize().
+  std::optional<Isolate> isolate_;
+
+  // Messages received before the controller is initialized from dart
+  // code. Buffered until Initialize() is called.
+  std::vector<WindowsMessage> pending_messages_;
```

Comment:
@mattkae I believe the conclusion from [this conversation](https://github.com/flutter/flutter/pull/168728/#discussion_r2152072022) was that we should remove pending messages. Please let me know if I missed something though :)